test/exe/main_common_test.cc

```diff
@@ -226,32 +149,187 @@ TEST_F(MainCommonTest, LegacyMain) {
+  bool waitForEnvoyToExit() {
+    finished_.WaitForNotification();
+    envoy_thread_->join();
+    return envoy_return_;
+  }
+
+  std::unique_ptr<Thread::Thread> envoy_thread_;
+  std::unique_ptr<MainCommon> main_common_;
+  absl::Notification started_;
+  absl::Notification finished_;
+  absl::Notification resume_;
+  absl::Notification pause_point_;
+  bool envoy_return_;
+  bool envoy_started_;
+  bool envoy_finished_;
+  bool pause_before_run_;
+  bool pause_after_run_;
+};
+
+TEST_F(AdminRequestTest, AdminRequestGetStatsAndQuit) {
   if (!Envoy::TestEnvironment::shouldRunTestForIpVersion(Network::Address::IpVersion::v4)) {
     return;
   }
-  addArg("--disable-hot-restart");
   startEnvoy();
-  waitForEnvoyToStart();
+  started_.WaitForNotification();
   EXPECT_THAT(adminRequest("/stats", "GET"), HasSubstr("filesystem.reopen_failed"));
   adminRequest("/quitquitquit", "POST");
   EXPECT_TRUE(waitForEnvoyToExit());
 }
 
 // This test is identical to the above one, except that instead of using an admin /quitquitquit,
 // we send ourselves a SIGTERM, which should have the same effect.
-TEST_F(MainCommonTest, AdminRequestGetStatsAndKill) {
+TEST_F(AdminRequestTest, AdminRequestGetStatsAndKill) {
   if (!Envoy::TestEnvironment::shouldRunTestForIpVersion(Network::Address::IpVersion::v4)) {
     return;
   }
-  addArg("--disable-hot-restart");
   startEnvoy();
-  waitForEnvoyToStart();
+  started_.WaitForNotification();
   EXPECT_THAT(adminRequest("/stats", "GET"), HasSubstr("filesystem.reopen_failed"));
   kill(getpid(), SIGTERM);
   EXPECT_TRUE(waitForEnvoyToExit());
 }
 
+TEST_F(AdminRequestTest, AdminRequestBeforeRun) {
+  if (!Envoy::TestEnvironment::shouldRunTestForIpVersion(Network::Address::IpVersion::v4)) {
+    return;
+  }
+  // Induce the situation where the Envoy thread is active, and main_common_ is constructed,
+  // but run() hasn't been issued yet. AdminRequests will not finish immediately, but will
+  // do so at some point after run() is allowed to start.
+  pause_before_run_ = true;
+  startEnvoy();
+  pause_point_.WaitForNotification();
+
+  bool admin_handler_was_called = false;
+  std::string out;
+  main_common_->adminRequest(
+      "/stats", "GET",
+      [&admin_handler_was_called, &out](const Http::HeaderMap& /*response_headers*/,
+                                        absl::string_view body) {
+        admin_handler_was_called = true;
+        out = std::string(body);
+      });
+
+  // The admin handler can't be called until after we let run() go.
+  EXPECT_FALSE(admin_handler_was_called);
+
+  // Now unblock the envoy thread so it can wake up and process outstanding posts.
+  resume_.Notify();
+
+  // We don't get a notification when run(), so it's not safe to check whether the
+  // admin handler is called until after we quit.
+  adminRequest("/quitquitquit", "POST");
+  EXPECT_TRUE(waitForEnvoyToExit());
+  EXPECT_TRUE(admin_handler_was_called);
+  EXPECT_THAT(out, HasSubstr("filesystem.reopen_failed"));
```

Comment:
Seems like a weird thing to look for in the output. What's trying to be reopened here and what's failed? Can you add a comment that points to the bit that logs that line?

Reply:
I was just checking for any stat.  Adding comment.